dmd.func: Return false in checkNRVO for all variables with nested refs

ibuclaw opened the pull request:
Whether it ends up in a closure, or stack frame.  The address of such variables cannot share the same address as the return slot/hidden pointer.

This removes the need to override isNRVO after semantic has finished.

ibuclaw committed:
dmd.func: Return false in checkNRVO for all variables with nested refs

kinke commented:
I've just changed this in LDC to work with stack frames, capturing the NRVO var by ref in that case.

ibuclaw commented:
> I've just changed this in LDC to work with stack frames, capturing the NRVO var by ref in that case.
> 
> **Edit**: test case: https://github.com/ldc-developers/ldc/blob/master/tests/codegen/nested_nrvo_gh3883.d

That is indeed surprising behaviour, I've adjusted the condition and added the test anyway.

kinke commented:
I noticed the ugliness in the glue code back then, but wasn't sure how to get rid of it and at what point calling `needsClosure()` is fine. So, glad you took care of it. :+1:
